File: design-pattern/builder-pattern/simple-builder/simple-builder.ts
class HttpRequest {
    public url: string = "";
    public method: string = "";
    public header: Map<string, string> = new Map();
    public timeOut: number = 0;

    public request(): void {
        console.log(
            "Request accepted with URL: " +
            this.url +
            " and Method: " +
            this.method
        );
    }
}

class HttpReqBuilder {
    private req: HttpRequest;

    constructor() {
        this.req = new HttpRequest();
    }

    setUrl(url: string): HttpReqBuilder {
        this.req.url = url;
        return this;
    }

    setMethod(method: string): HttpReqBuilder {
        this.req.method = method;
        return this;
    }

    setTimeOut(timer: number): HttpReqBuilder {
        this.req.timeOut = timer;
        return this;
    }

    build(): HttpRequest {
        // validation
        if (!this.req.url) {
            throw new Error("URL not found");
        }
        return this.req;
    }
}

// Usage
const newReq = new HttpReqBuilder()
    .setUrl("www.omkar.com")
    .setMethod("GET")
    .setTimeOut(1000)
    .build();

newReq.request();

File: design-pattern/builder-pattern/simple-builder/simple-builder.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std;

class HttpsReq
{
private:
    string url;
    string method;
    map<string, string> header;
    map<string, string> query;
    int timeout;

public:
    friend class HttpReqBuilder;

    void request()
    {
        cout << "Request is processing with " << url << " : " << method << endl;
    }
};

class HttpReqBuilder
{
private:
    HttpsReq *req;

public:
    // Constructor to initialize req
    HttpReqBuilder()
    {
        req = new HttpsReq();
    }

    HttpReqBuilder &getUrl(string url)
    {
        req->url = url;
        return *this;
    }

    HttpReqBuilder &getMethod(string method)
    {
        req->method = method;
        return *this;
    }

    HttpReqBuilder &getHeader(string key, string value)
    {
        req->header[key] = value;
        return *this;
    }

    HttpReqBuilder &getQuery(string key, string value)
    {
        req->query[key] = value;
        return *this;
    }

    HttpReqBuilder &getTime(int time)
    {
        req->timeout = time;
        return *this;
    }

    HttpsReq *build()
    {
        if (req->url.empty())
        {
            throw runtime_error("URL cannot be empty");
        }
        return req;
    }
};

int main()
{
    HttpsReq *newReq = HttpReqBuilder()
                           .getUrl("www.omkar.com")
                           .getMethod("GET")
                           .getTime(1000)
                           .build();

    newReq->request();

    return 0;
}